File: src/routes/auto_download.py
"""Auto download API routes."""
import logging
import threading
from pathlib import Path
from typing import Optional

from flask import Blueprint, request

from src.constants.api import success_response, error_response
from src.services.auto_download_service import (
    get_auto_download_service,
    DownloadConfig,
    DownloadFilter,
    DownloadStatus,
)
from src.services.video_downloader import get_video_downloader

logger = logging.getLogger(__name__)

auto_download_bp = Blueprint("auto_download", __name__, url_prefix="/api/auto-download")

# Global download manager
_download_manager: Optional["DownloadManager"] = None


class DownloadManager:
    """Manages active downloads and coordinates between service and downloader."""

    def __init__(self):
        self.auto_service = get_auto_download_service()
        self.downloader: Optional[object] = None
        self._lock = threading.Lock()
        self._active_downloads: set = set()

    def get_downloader(self):
        """Get or create video downloader."""
        if self.downloader is None:
            download_path = self.auto_service.get_download_path()
            config = self.auto_service.get_config()
            self.downloader = get_video_downloader(
                download_path, config.max_concurrent_downloads
            )
            # Register progress callback
            self.downloader.register_progress_callback(self._on_progress_update)
        return self.downloader

    def _on_progress_update(self, progress):
        """Handle download progress updates."""
        # Update record in auto download service
        from src.services.auto_download_service import DownloadRecord

        if progress.status in ["completed", "failed", "cancelled"]:
            with self._lock:
                self._active_downloads.discard(progress.episode_id)

        # Find and update record
        record = DownloadRecord(
            anime_id="",  # Will be filled from existing record
            anime_title="",
            episode_id=progress.episode_id,
            episode_num="",
            status=DownloadStatus(progress.status),
            file_path=progress.file_path,
            error_message=progress.error_message,
        )
        self.auto_service.update_download_record(record)

    def start_download(self, anime: dict, episode: dict) -> dict:
        """Start a new download.

        Args:
            anime: Anime dictionary
            episode: Episode dictionary

        Returns:
            Result dictionary
        """
        episode_id = episode["id"]

        # Check if already downloading
        with self._lock:
            if episode_id in self._active_downloads:
                return {"success": False, "message": "Already downloading"}
            self._active_downloads.add(episode_id)

        # Create download record
        from src.services.auto_download_service import DownloadRecord, DownloadStatus

        record = DownloadRecord(
            anime_id=anime["id"],
            anime_title=anime["title"],
            episode_id=episode_id,
            episode_num=episode["episode"],
            status=DownloadStatus.DOWNLOADING,
        )
        self.auto_service.add_download_record(record)

        # Start download in background
        def do_download():
            downloader = self.get_downloader()
            episode_url = episode.get("url", f"https://anime1.me/{episode_id}")
            downloader.download(
                episode_id=episode_id,
                episode_url=episode_url,
                anime_title=anime["title"],
                episode_num=episode["episode"],
            )

        thread = threading.Thread(target=do_download, daemon=True)
        thread.start()

        return {"success": True, "message": "Download started"}

    def get_progress(self, episode_id: Optional[str] = None) -> dict:
        """Get download progress.

        Args:
            episode_id: Optional episode ID to get specific progress

        Returns:
            Progress information
        """
        if self.downloader is None:
            return {"downloads": []}

        if episode_id:
            progress = self.downloader.get_progress(episode_id)
            if progress:
                return {"download": progress.to_dict()}
            return {"error": "Download not found"}

        all_progress = self.downloader.get_all_progress()
        return {
            "downloads": [p.to_dict() for p in all_progress.values()],
            "active_count": len(self._active_downloads),
        }


def get_download_manager() -> DownloadManager:
    """Get or create the global download manager."""
    global _download_manager
    if _download_manager is None:
        _download_manager = DownloadManager()
    return _download_manager


@auto_download_bp.route("/config", methods=["GET"])
def get_config():
    """Get auto download configuration.

    Returns:
        Current configuration including filters and settings.
    """
    try:
        service = get_auto_download_service()
        config = service.get_config()

        return success_response(
            data={
                "enabled": config.enabled,
                "download_path": config.download_path,
                "check_interval_hours": config.check_interval_hours,
                "max_concurrent_downloads": config.max_concurrent_downloads,
                "auto_download_new": config.auto_download_new,
                "auto_download_favorites": config.auto_download_favorites,
                "filters": {
                    "min_year": config.filters.min_year,
                    "max_year": config.filters.max_year,
                    "specific_years": config.filters.specific_years,
                    "seasons": config.filters.seasons,
                    "min_episodes": config.filters.min_episodes,
                    "include_patterns": config.filters.include_patterns,
                    "exclude_patterns": config.filters.exclude_patterns,
                },
            }
        )
    except Exception as e:
        logger.error(f"Error getting auto download config: {e}")
        return error_response(str(e), 500)


@auto_download_bp.route("/config", methods=["POST"])
def update_config():
    """Update auto download configuration.

    Request body:
        enabled: bool - Enable/disable auto download
        download_path: str - Custom download path (empty for default)
        check_interval_hours: int - Check interval in hours
        max_concurrent_downloads: int - Max concurrent downloads
        auto_download_new: bool - Auto download new anime
        auto_download_favorites: bool - Auto download favorites
        filters: object - Filter configuration
            min_year: int - Minimum year
            max_year: int - Maximum year
            specific_years: list - Specific years to download
            seasons: list - Seasons to download ["冬季", "春季", "夏季", "秋季"]
            min_episodes: int - Minimum episode count
            include_patterns: list - Regex patterns to include
            exclude_patterns: list - Regex patterns to exclude

    Returns:
        Updated configuration
    """
    try:
        data = request.get_json() or {}
        service = get_auto_download_service()

        # Get current config
        config = service.get_config()

        # Update fields
        if "enabled" in data:
            config.enabled = bool(data["enabled"])
        if "download_path" in data:
            config.download_path = str(data["download_path"])
        if "check_interval_hours" in data:
            config.check_interval_hours = int(data["check_interval_hours"])
        if "max_concurrent_downloads" in data:
            config.max_concurrent_downloads = int(data["max_concurrent_downloads"])
        if "auto_download_new" in data:
            config.auto_download_new = bool(data["auto_download_new"])
        if "auto_download_favorites" in data:
            config.auto_download_favorites = bool(data["auto_download_favorites"])

        # Update filters
        if "filters" in data:
            filters_data = data["filters"]
            if "min_year" in filters_data:
                config.filters.min_year = (
                    int(filters_data["min_year"]) if filters_data["min_year"] else None
                )
            if "max_year" in filters_data:
                config.filters.max_year = (
                    int(filters_data["max_year"]) if filters_data["max_year"] else None
                )
            if "specific_years" in filters_data:
                config.filters.specific_years = [
                    int(y) for y in filters_data["specific_years"]
                ]
            if "seasons" in filters_data:
                config.filters.seasons = list(filters_data["seasons"])
            if "min_episodes" in filters_data:
                config.filters.min_episodes = (
                    int(filters_data["min_episodes"])
                    if filters_data["min_episodes"]
                    else None
                )
            if "include_patterns" in filters_data:
                config.filters.include_patterns = list(filters_data["include_patterns"])
            if "exclude_patterns" in filters_data:
                config.filters.exclude_patterns = list(filters_data["exclude_patterns"])

        # Save config
        if service.update_config(config):
            # Restart scheduler if enabled changed
            if config.enabled:
                service.stop_scheduler()
                service.start_scheduler()
            else:
                service.stop_scheduler()

            return success_response(
                message="Configuration updated",
                data=config.to_dict(),
            )
        else:
            return error_response("Failed to save configuration", 500)

    except Exception as e:
        logger.error(f"Error updating auto download config: {e}")
        return error_response(str(e), 500)


@auto_download_bp.route("/status", methods=["GET"])
def get_status():
    """Get auto download service status.

    Returns:
        Service status including recent downloads and counts.
    """
    try:
        service = get_auto_download_service()
        status = service.get_status()
        return success_response(data=status)
    except Exception as e:
        logger.error(f"Error getting auto download status: {e}")
        return error_response(str(e), 500)


@auto_download_bp.route("/history", methods=["GET"])
def get_history():
    """Get download history.

    Query params:
        limit: Maximum number of records (default: 100)
        status: Filter by status (pending, downloading, completed, failed, skipped)

    Returns:
        List of download records
    """
    try:
        service = get_auto_download_service()
        limit = request.args.get("limit", 100, type=int)
        status_str = request.args.get("status", None)

        status = DownloadStatus(status_str) if status_str else None
        history = service.get_history(limit=limit, status=status)

        return success_response(
            data={
                "history": [r.to_dict() for r in history],
                "total": len(history),
            }
        )
    except Exception as e:
        logger.error(f"Error getting download history: {e}")
        return error_response(str(e), 500)


@auto_download_bp.route("/download", methods=["POST"])
def start_download():
    """Start a manual download.

    Request body:
        anime_id: str - Anime ID
        episode_id: str - Episode ID
        episode_num: str - Episode number
        title: str - Anime title
        url: str - Episode URL (optional)

    Returns:
        Download start result
    """
    try:
        data = request.get_json() or {}

        anime_id = data.get("anime_id")
        episode_id = data.get("episode_id")
        episode_num = data.get("episode_num")
        title = data.get("title")
        url = data.get("url", f"https://anime1.me/{episode_id}")

        if not all([anime_id, episode_id, title]):
            return error_response("Missing required fields: anime_id, episode_id, title", 400)

        anime = {
            "id": anime_id,
            "title": title,
        }
        episode = {
            "id": episode_id,
            "episode": episode_num or "",
            "url": url,
        }

        manager = get_download_manager()
        result = manager.start_download(anime, episode)

        if result["success"]:
            return success_response(message=result["message"], data=result)
        else:
            return error_response(result["message"], 400)

    except Exception as e:
        logger.error(f"Error starting download: {e}")
        return error_response(str(e), 500)


@auto_download_bp.route("/progress", methods=["GET"])
def get_progress():
    """Get download progress.

    Query params:
        episode_id: Specific episode ID (optional, returns all if not specified)

    Returns:
        Download progress information
    """
    try:
        episode_id = request.args.get("episode_id")
        manager = get_download_manager()
        progress = manager.get_progress(episode_id)
        return success_response(data=progress)
    except Exception as e:
        logger.error(f"Error getting download progress: {e}")
        return error_response(str(e), 500)


@auto_download_bp.route("/filter/preview", methods=["POST"])
def preview_filter():
    """Preview which anime would match the given filter.

    Request body:
        filters: Filter configuration object

    Returns:
        List of anime that would match the filter
    """
    try:
        data = request.get_json() or {}
        filters_data = data.get("filters", {})

        # Create temporary filter
        temp_filter = DownloadFilter()
        if "min_year" in filters_data:
            temp_filter.min_year = (
                int(filters_data["min_year"]) if filters_data["min_year"] else None
            )
        if "max_year" in filters_data:
            temp_filter.max_year = (
                int(filters_data["max_year"]) if filters_data["max_year"] else None
            )
        if "specific_years" in filters_data:
            temp_filter.specific_years = [int(y) for y in filters_data["specific_years"]]
        if "seasons" in filters_data:
            temp_filter.seasons = list(filters_data["seasons"])
        if "min_episodes" in filters_data:
            temp_filter.min_episodes = (
                int(filters_data["min_episodes"]) if filters_data["min_episodes"] else None
            )
        if "include_patterns" in filters_data:
            temp_filter.include_patterns = list(filters_data["include_patterns"])
        if "exclude_patterns" in filters_data:
            temp_filter.exclude_patterns = list(filters_data["exclude_patterns"])

        # Get all anime and filter
        from src.services.anime_cache_service import get_anime_list_cache

        all_anime = get_anime_list_cache()
        all_anime_dicts = [a.to_dict() for a in all_anime]

        service = get_auto_download_service()
        filtered = service.filter_anime(all_anime_dicts)

        # Also apply temporary filter
        final_filtered = [a for a in filtered if temp_filter.matches(a)]

        return success_response(
            data={
                "total_anime": len(all_anime_dicts),
                "matched_count": len(final_filtered),
                "matched_anime": final_filtered[:50],  # Limit to 50 for preview
            }
        )

    except Exception as e:
        logger.error(f"Error previewing filter: {e}")
        return error_response(str(e), 500)


@auto_download_bp.route("/test", methods=["POST"])
def test_download():
    """Test download a single episode.

    Request body:
        episode_url: URL of the episode to test

    Returns:
        Test result with video info
    """
    try:
        data = request.get_json() or {}
        episode_url = data.get("episode_url")

        if not episode_url:
            return error_response("episode_url is required", 400)

        # Create temporary downloader to test
        service = get_auto_download_service()
        download_path = service.get_download_path()

        from src.services.video_downloader import VideoDownloader

        downloader = VideoDownloader(download_path, max_concurrent=1)

        try:
            # Get video info
            video_info = downloader._get_video_info(episode_url)
            if not video_info:
                return error_response("Could not extract video information", 400)

            # Get video URL
            video_url = downloader._get_video_url(video_info.get("vid", ""))

            return success_response(
                data={
                    "success": True,
                    "video_info": video_info,
                    "video_url_found": video_url is not None,
                    "message": "Video info extracted successfully",
                }
            )
        finally:
            downloader.close()

    except Exception as e:
        logger.error(f"Error testing download: {e}")
        return error_response(str(e), 500)

File: src/services/autoDownload.js
/**
 * Auto Download Service
 * 
 * Features:
 * - Download configuration with filters
 * - Download task management with VideoDownloader
 * - Download history tracking
 * - Background scheduler (Node.js setInterval)
 */

import { app } from 'electron';
import path from 'path';
import fs from 'fs';
import { settingsDB } from './database.js';
import { getVideoDownloader, DownloadStatus } from './videoDownloader.js';
import { getVideoInfo } from './videoProxy.js';

// Export DownloadStatus for compatibility
export { DownloadStatus };

// Default download config
const DEFAULT_CONFIG = {
  enabled: false,
  downloadPath: '',
  checkIntervalHours: 24,
  maxConcurrentDownloads: 2,
  autoDownloadNew: false,
  autoDownloadFavorites: true,
  filters: {
    minYear: null,
    maxYear: null,
    specificYears: [],
    seasons: [], // ['冬季', '春季', '夏季', '秋季']
    minEpisodes: null,
    includePatterns: [],
    excludePatterns: [],
  }
};

class AutoDownloadService {
  constructor() {
    this._config = null;
    this._history = [];
    this._schedulerInterval = null;
    this._isRunning = false;
    this._downloadCallbacks = [];
    this._videoDownloader = null;
    this._webContents = null; // For sending events to renderer
  }

  /**
   * Set WebContents for sending events to renderer
   */
  setWebContents(webContents) {
    this._webContents = webContents;
  }

  /**
   * Initialize the service
   */
  async init() {
    await this._loadConfig();
    await this._loadHistory();
    
    // Ensure config is initialized
    if (!this._config) {
      this._config = { ...DEFAULT_CONFIG };
    }
    
    // Initialize video downloader
    this._initVideoDownloader();
    
    // Start scheduler if enabled
    if (this._config.enabled) {
      this.startScheduler();
    }
  }

  /**
   * Initialize video downloader
   */
  _initVideoDownloader() {
    const downloadPath = this.getDownloadPath();
    const maxConcurrent = this._config?.maxConcurrentDownloads || 2;
    
    this._videoDownloader = getVideoDownloader(downloadPath, maxConcurrent);
    
    // Register event handlers
    this._videoDownloader.on('started', (progress) => {
      this._updateDownloadRecord(progress.episodeId, {
        status: DownloadStatus.DOWNLOADING,
        progress: 0,
      });
      this._notifyRenderer('download-started', progress);
    });
    
    this._videoDownloader.on('progress', (progress) => {
      this._updateDownloadRecord(progress.episodeId, {
        status: DownloadStatus.DOWNLOADING,
        progress: progress.percent,
        totalBytes: progress.totalBytes,
        downloadedBytes: progress.downloadedBytes,
        speedBytesPerSec: progress.speedBytesPerSec,
      });
      this._notifyRenderer('download-progress', progress);
    });
    
    this._videoDownloader.on('completed', (progress) => {
      this._updateDownloadRecord(progress.episodeId, {
        status: DownloadStatus.COMPLETED,
        progress: 100,
        filePath: progress.filePath,
        completedAt: new Date().toISOString(),
      });
      this._notifyRenderer('download-completed', progress);
    });
    
    this._videoDownloader.on('error', (progress) => {
      this._updateDownloadRecord(progress.episodeId, {
        status: DownloadStatus.FAILED,
        errorMessage: progress.errorMessage,
        completedAt: new Date().toISOString(),
      });
      this._notifyRenderer('download-error', progress);
    });
    
    this._videoDownloader.on('cancelled', (progress) => {
      this._updateDownloadRecord(progress.episodeId, {
        status: DownloadStatus.CANCELLED,
        completedAt: new Date().toISOString(),
      });
      this._notifyRenderer('download-cancelled', progress);
    });
  }

  /**
   * Notify renderer process of download events
   */
  _notifyRenderer(event, data) {
    // Call registered callbacks
    for (const callback of this._downloadCallbacks) {
      try {
        callback(event, data);
      } catch (error) {
        console.error('[AutoDownload] Callback error:', error);
      }
    }
    
    // Send to renderer via WebContents
    if (this._webContents && !this._webContents.isDestroyed()) {
      this._webContents.send('auto-download:' + event, data);
    }
  }

  /**
   * Update download record in history
   */
  _updateDownloadRecord(episodeId, updates) {
    const record = this._history.find(r => r.episodeId === episodeId);
    if (record) {
      Object.assign(record, updates);
      this._saveHistory();
    }
  }

  /**
   * Get download directory
   */
  getDownloadPath() {
    const config = this._config || DEFAULT_CONFIG;
    if (config.downloadPath) {
      const dir = config.downloadPath;
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      return dir;
    }
    
    // Default to Downloads/Anime1
    const downloadsPath = path.join(app.getPath('downloads'), 'Anime1');
    if (!fs.existsSync(downloadsPath)) {
      fs.mkdirSync(downloadsPath, { recursive: true });
    }
    return downloadsPath;
  }

  /**
   * Load config from database
   */
  async _loadConfig() {
    try {
      const saved = await settingsDB.get('autoDownloadConfig');
      this._config = saved ? { ...DEFAULT_CONFIG, ...saved } : { ...DEFAULT_CONFIG };
    } catch (error) {
      console.error('[AutoDownload] Error loading config:', error);
      this._config = { ...DEFAULT_CONFIG };
    }
  }

  /**
   * Save config to database
   */
  async _saveConfig() {
    try {
      await settingsDB.set('autoDownloadConfig', this._config);
    } catch (error) {
      console.error('[AutoDownload] Error saving config:', error);
    }
  }

  /**
   * Load download history from database
   */
  async _loadHistory() {
    try {
      const saved = await settingsDB.get('autoDownloadHistory');
      this._history = saved || [];
    } catch (error) {
      console.error('[AutoDownload] Error loading history:', error);
      this._history = [];
    }
  }

  /**
   * Save download history to database
   */
  async _saveHistory() {
    try {
      // Keep only last 1000 records
      const trimmedHistory = this._history.slice(-1000);
      await settingsDB.set('autoDownloadHistory', trimmedHistory);
    } catch (error) {
      console.error('[AutoDownload] Error saving history:', error);
    }
  }

  /**
   * Get current config
   */
  getConfig() {
    return { ...(this._config || DEFAULT_CONFIG) };
  }

  /**
   * Update config
   */
  async updateConfig(newConfig) {
    this._config = { ...(this._config || DEFAULT_CONFIG), ...newConfig };
    await this._saveConfig();
    
    // Update video downloader settings
    if (this._videoDownloader) {
      // Recreate with new settings if max concurrent changed
      const maxConcurrent = this._config.maxConcurrentDownloads || 2;
      // Note: maxConcurrent is set at creation, would need to recreate to change
    }
    
    // Restart scheduler if running state changed
    const config = this._config || DEFAULT_CONFIG;
    if (config.enabled && !this._isRunning) {
      this.startScheduler();
    } else if (!config.enabled && this._isRunning) {
      this.stopScheduler();
    }
    
    return true;
  }

  /**
   * Get download history
   */
  getHistory(limit = 50, status = null) {
    let history = [...this._history];
    
    if (status) {
      history = history.filter(r => r.status === status);
    }
    
    // Sort by createdAt desc
    history.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
    
    return history.slice(0, limit);
  }
  
  /**
   * Get all downloads
   */
  getAllDownloads() {
    return [...this._history];
  }

  /**
   * Get active downloads
   */
  getActiveDownloads() {
    if (!this._videoDownloader) return [];
    return this._videoDownloader.getActiveDownloads();
  }

  /**
   * Check if anime matches filters
   */
  matchesFilters(anime) {
    if (!this._config || !this._config.filters) {
      return true; // If config not loaded, allow all
    }
    const filters = this._config.filters;
    
    // Year filter
    if (anime.year && filters.minYear) {
      const year = parseInt(anime.year);
      if (!isNaN(year) && year < filters.minYear) return false;
    }
    if (anime.year && filters.maxYear) {
      const year = parseInt(anime.year);
      if (!isNaN(year) && year > filters.maxYear) return false;
    }
    if (filters.specificYears?.length > 0 && anime.year) {
      const year = parseInt(anime.year);
      if (!filters.specificYears.includes(year)) return false;
    }
    
    // Season filter
    if (filters.seasons?.length > 0 && anime.season) {
      if (!filters.seasons.includes(anime.season)) return false;
    }
    
    // Episode filter
    if (filters.minEpisodes && anime.episode) {
      const ep = parseInt(anime.episode);
      if (!isNaN(ep) && ep < filters.minEpisodes) return false;
    }
    
    // Include patterns
    if (filters.includePatterns?.length > 0 && anime.title) {
      const matched = filters.includePatterns.some(pattern => {
        try {
          const regex = new RegExp(pattern, 'i');
          return regex.test(anime.title);
        } catch {
          return anime.title.includes(pattern);
        }
      });
      if (!matched) return false;
    }
    
    // Exclude patterns
    if (filters.excludePatterns?.length > 0 && anime.title) {
      const excluded = filters.excludePatterns.some(pattern => {
        try {
          const regex = new RegExp(pattern, 'i');
          return regex.test(anime.title);
        } catch {
          return anime.title.includes(pattern);
        }
      });
      if (excluded) return false;
    }
    
    return true;
  }

  /**
   * Start the background scheduler
   */
  startScheduler() {
    if (this._isRunning) {
      console.log('[AutoDownload] Scheduler already running');
      return;
    }
    
    const config = this._config || DEFAULT_CONFIG;
    if (!config.enabled) {
      console.log('[AutoDownload] Not starting scheduler (disabled)');
      return;
    }
    
    this._isRunning = true;
    console.log('[AutoDownload] Scheduler started');
    
    // Run immediately, then schedule
    this._checkAndDownload();
    
    // Schedule next check
    const intervalMs = config.checkIntervalHours * 60 * 60 * 1000;
    this._schedulerInterval = setInterval(() => {
      this._checkAndDownload();
    }, intervalMs);
  }

  /**
   * Stop the scheduler
   */
  stopScheduler() {
    if (!this._isRunning) return;
    
    if (this._schedulerInterval) {
      clearInterval(this._schedulerInterval);
      this._schedulerInterval = null;
    }
    
    this._isRunning = false;
    console.log('[AutoDownload] Scheduler stopped');
  }

  /**
   * Check for new episodes and download
   */
  async _checkAndDownload() {
    console.log('[AutoDownload] Running scheduled check...');
    
    try {
      // This will be implemented to check favorites and new anime
      // For now, just log that check ran
      console.log('[AutoDownload] Check completed');
    } catch (error) {
      console.error('[AutoDownload] Error during check:', error);
    }
  }

  /**
   * Start a new download
   */
  async startDownload(anime, episode, videoUrl = null, cookies = null) {
    const episodeId = episode.id;
    
    // Check if already downloaded
    const existing = this._history.find(r => 
      r.episodeId === episodeId && r.status === DownloadStatus.COMPLETED
    );
    if (existing) {
      throw new Error('Episode already downloaded');
    }

    // If no videoUrl provided, fetch it
    let finalVideoUrl = videoUrl;
    let finalCookies = cookies || {};
    
    if (!finalVideoUrl) {
      const videoInfo = await getVideoInfo(episode.url);
      if (!videoInfo.success) {
        throw new Error(videoInfo.error || 'Failed to get video info');
      }
      finalVideoUrl = videoInfo.url;
      finalCookies = videoInfo.cookies || {};
    }

    // Create download record
    const record = {
      id: `${anime.id}_${episodeId}`,
      animeId: anime.id,
      animeTitle: anime.title,
      episodeId: episodeId,
      episodeNum: episode.episode || episode.title,
      status: DownloadStatus.PENDING,
      progress: 0,
      totalBytes: 0,
      downloadedBytes: 0,
      filePath: '',
      errorMessage: '',
      createdAt: new Date().toISOString(),
      completedAt: null,
    };
    
    this._history.push(record);
    await this._saveHistory();
    
    // Add to video downloader
    try {
      const progress = await this._videoDownloader.addDownload(
        anime,
        episode,
        finalVideoUrl,
        finalCookies
      );
      
      // Update record with initial progress
      Object.assign(record, {
        status: progress.status,
        totalBytes: progress.totalBytes,
        downloadedBytes: progress.downloadedBytes,
      });
      await this._saveHistory();
      
      return record;
    } catch (error) {
      // Update record with error
      record.status = DownloadStatus.FAILED;
      record.errorMessage = error.message;
      record.completedAt = new Date().toISOString();
      await this._saveHistory();
      throw error;
    }
  }

  /**
   * Cancel a download
   */
  async cancelDownload(episodeId) {
    if (!this._videoDownloader) {
      throw new Error('Video downloader not initialized');
    }
    
    const result = this._videoDownloader.cancelDownload(episodeId);
    
    // Update record
    const record = this._history.find(r => r.episodeId === episodeId);
    if (record) {
      record.status = DownloadStatus.CANCELLED;
      record.completedAt = new Date().toISOString();
      await this._saveHistory();
    }
    
    return result;
  }

  /**
   * Delete a download record and file
   */
  async deleteDownload(downloadId) {
    const index = this._history.findIndex(r => r.id === downloadId);
    if (index === -1) {
      throw new Error('Download record not found');
    }
    
    const record = this._history[index];
    
    // Delete file if exists
    if (record.filePath && fs.existsSync(record.filePath)) {
      try {
        fs.unlinkSync(record.filePath);
      } catch (error) {
        console.error('[AutoDownload] Error deleting file:', error);
      }
    }
    
    // Remove from history
    this._history.splice(index, 1);
    await this._saveHistory();
    
    return true;
  }

  /**
   * Get service status
   */
  getStatus() {
    const statusCounts = {
      pending: 0,
      downloading: 0,
      completed: 0,
      failed: 0,
      cancelled: 0,
    };
    
    for (const record of this._history) {
      statusCounts[record.status] = (statusCounts[record.status] || 0) + 1;
    }
    
    const config = this._config || DEFAULT_CONFIG;
    
    return {
      enabled: config.enabled,
      running: this._isRunning,
      downloadPath: this.getDownloadPath(),
      checkIntervalHours: config.checkIntervalHours,
      maxConcurrentDownloads: config.maxConcurrentDownloads,
      activeDownloads: this.getActiveDownloads().length,
      statusCounts,
      recentDownloads: this.getHistory(10),
    };
  }

  /**
   * Preview filter results on anime list
   */
  previewFilter(animeList) {
    return animeList.filter(anime => this.matchesFilters(anime));
  }

  /**
   * Register download callback
   */
  onDownload(callback) {
    this._downloadCallbacks.push(callback);
  }

  /**
   * Unregister download callback
   */
  offDownload(callback) {
    const index = this._downloadCallbacks.indexOf(callback);
    if (index > -1) {
      this._downloadCallbacks.splice(index, 1);
    }
  }

  /**
   * Clear completed downloads from memory
   */
  clearCompletedDownloads() {
    if (this._videoDownloader) {
      this._videoDownloader.clearCompleted();
    }
  }

  /**
   * Destroy service
   */
  destroy() {
    this.stopScheduler();
    
    if (this._videoDownloader) {
      this._videoDownloader.destroy();
      this._videoDownloader = null;
    }
    
    this._downloadCallbacks = [];
  }
}

// Singleton instance
let autoDownloadService = null;

export function getAutoDownloadService() {
  if (!autoDownloadService) {
    autoDownloadService = new AutoDownloadService();
  }
  return autoDownloadService;
}

export function resetAutoDownloadService() {
  if (autoDownloadService) {
    autoDownloadService.destroy();
    autoDownloadService = null;
  }
}

export default {
  getAutoDownloadService,
  resetAutoDownloadService,
  DownloadStatus,
};
